docs/exernal-services.md

```diff
@@ -0,0 +1,206 @@
+
+When the ActivityRuntimeAPI is used to create or update an ExternalActivity, the `launch_url` is only set when it is created. Because of this the duplicate_on_lara code sets the `launch_url` first before doing the internal publish.
+
+When data_helpers.rb generates JSON from the activity it includes lara_activity_or_sequence which is based on whether the activity has `launch_url`. This seems be used to determine which links are shown to the user.
+
+## Client
+Clients are used for authentication. When the Portal creates an AccessGrant to provide a token to an external service, it generally has an associated Client. The Client is used to verify requests with the token. If the client specifies domain_matchers then only requests from those domains are accepted.
+
+When the request is part of an OAuth2 authorize, then the Client needs to specify the exact redirect that the Portal is sending this token or code back to. These OAuth2 authorize requests are used when LARA is doing OAuth2. They are also used by the single page apps that use the Portal for authentication like the token-service, glossary authoring, or a portal-report when not launched directly from the portal.
+
+## Seeding the Database
+To make it easier for developers, when the database is seeded a few of the resources above are automatically created.
+
+An ExternalReport is created for the portal-report. And a Client is created that this ExternalReport uses.
+
+# Ways to Improve
+
+## Eliminate the source_type
+
+### default_report
+The use of source_type to connect ExternalActivities to default reports is not needed. Instead a 'default_report' can be added to the Tool model. This will have the same behavior and simplify the modeling.
+
+### remove filtering of runnables in user researcher report
+Currently the runnables are filtered by source_type in the user researcher report page. This is not really useful, and seems could just be removed. If we need filtering like this, we could add Tool filtering
+
+### duplicate activity support
+Currently a source_type of LARA means the portal uses the `remote_duplicate` action in LARA when an ExternalActivity is duplicated. It is convenient for authors so they can duplicate resources in the portal without need to setup the ExternalActivity for the resource (description, grade_levels, ...). Instead of using this check on a source_type of LARA, this should be made generic so non-LARA authoring systems can provide support for duplicating from the Portal.
```

Comment:
```suggestion
Currently a `source_type` of LARA means the portal uses the `remote_duplicate` action in LARA when an ExternalActivity is duplicated. It is convenient for authors so they can duplicate resources in the portal without needing to set up the ExternalActivity for the resource (description, grade_levels, ...). Instead of using this check on a `source_type` of LARA, this should be made generic so non-LARA authoring systems can provide support for duplicating from the Portal.
```